### DISCUSSION TITLE
Auto Materialization - Cross Code locations

### DISCUSSION THREAD

**Discussion by cody-scott** [dagster-io/dagster#19263] [Q&A] [answered]
Hi, 

Two questions, do auto materialization policies support cross code locations? Our DBT project lives in its own code location, so, ideally, if an auto materialization could trigger off the models in a different location it would save lots of `source` style tables in the DBT project.

Another question as well, what is the status on auto materializations experimental state; is it planned on becoming stable, or is it still undecided at this stage for the future support?

**Comment by elijahproffitt-2b** [comment, marked as answer]
I worked on this exact thing today. I do not work for dagster but hopefully this is helpful / someone can confirm or deny if this is a good approach.

I was able to define cross-code-location asset dependencies for asset decorators using the information in this doc: https://docs.dagster.io/concepts/assets/software-defined-assets#defining-external-asset-dependencies

For graph_asset decorators, those do not have a "deps" parameter. However, I was able to get this to work by doing something like this for the graph_asset:

```python
from dagster import graph_asset, AutoMaterializePolicy, AssetKey, AssetIn
from ..graphs import my_graph

@graph_asset(
    auto_materialize_policy=AutoMaterializePolicy.eager(),
    ins={
        "upstream_asset": AssetIn(
            key=AssetKey(["key_prefix","key"])
            )
    },
)
def downstream_asset(upstream_asset):
    return my_graph(upstream_asset)
```
this for the graph:
```python
from ..ops import my_op

@graph(
    ins={
        "upstream": In(Nothing) 
    }
)
def my_graph(upstream) -> Any:
    return my_op(upstream)
```
and then this for the op:
```python
@op
def my_op(upstream) -> Any:
    print(upstream)
    return 1
```

Lastly, in your code location's Definitions object, you use the "SourceAsset" to reference the upstream asset in the remote code location by its key_prefix / key values.

```python
from dagster import Definitions, SourceAsset, Assetkey
from .assets import downstream_asset

upstream_asset = SourceAsset(key=AssetKey(["key_prefix", "key"])

defs = Definitions(
    assets=[
        downstream_asset,
        upstream_asset,
    ]
)
```
This gave me the following capabilities.. I haven't tested this on kubernetes yet, but it seems to work locally when i run dagster dev with two separate code locations:
- auto-materialization of assets across code locations 
- access to the upstream asset's output (see print(upstream) in the "my_op")

Hopefully this helps.
